{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "convertible-indication",
   "metadata": {},
   "source": [
    "# Create a list of reading frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "motivated-storm",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "#listreadingframes:\n",
    "minlengthreadingframe=20\n",
    "maxlengthreadingframe=40\n",
    "\n",
    "filenamereadingframes=\"Output/sirinsp450permutationsof4/readingframes.txt\"\n",
    "listreadingframe=[]\n",
    "\n",
    "stopcodon1=('TGA', 'TAA', 'TAG')\n",
    "startcodon1=('ATG',\"GTG\")\n",
    "stopcodon2=(\"TCA\",\"TTA\",\"CTA\")\n",
    "startcodon2=(\"CAT\",\"CAC\")\n",
    "\n",
    "#source=https://parts.igem.org/Ribosome_Binding_Sites/Catalog ,http://arep.med.harvard.edu/labgc/adnan/projects/Utilities/revcomp.html\n",
    "ribosomalbindingsites1=[\"AAAGA[A-Z]{3}GA[A-Z]{3}\",\"AACTAGAATCACCTCTTGCTTTTGGGTAAGAAAGAGGAGA\",\"AACTAGAATCACCTCTTGGATTTGGGTATTAAAGAGGAGA\",\"ATTAAAGAGGAGAAA\",\"TTCACACAGGAAACC\",'GTGTG','GTGTGTCTAG','TCACACAGGAAACCGGTTCGATG','TCACACAGGAAAGGCCTCGATG','TCACACAGGACGGCCGGATG','TCTCACGTGTGTCAAG','TCTCACGTGTGT','CATCCCT','TCACATCCCT','TCACATCCCTCC','ACTGCACGAGGTAACACAAG','TACGAGGAGGATGAAGAGTA','ACTTTACTTATGAGGGAGTA','ACGAAGACGGAGACTTCTAA','AACCCTCAGGAGGTAAACCA','AAGACATGGAGACACATTTA','ACTGCACGAGGTAACACAAG','GAGGAGGATGAAGAGTAATGTGAAGAGCTG','AGGAGGTCATC','GCAAGCTCTTTTTTCAGTTGTCTC','CTGATAGTTAAAATCACCAGCATGA','TAAAAACAAGAGGAAAACAA','TCTCCTCTTT','ACGGAGAAGCAGCGAA','GAGGTTGGGACAAG','TAAATGTATCCGTTTATAAGGACAGCCCGA','CTCTTAAGTGGGAGCGGCT','CTCTACCGGAGAAATT','CTCATCGTTAAAGAGCGACTAC','CTCAGCCTGTACCTGGAGAGCCTTTC','CTCAAGGAGG','GAGAGG','AGGAGGATTACAA','AAAGAGGAGAAA','TCACACAGGAAAG','GGAAGAGG','TTTCTCCTCTTTAAT','TCACACAGGAAAGGCCTCG','ATTAAAGAGGAGAAATTAAGC',' TCGTTTCTGAAAAATTTTCGTTTCTGAAAA','TGGCTAACATAGGGT','TGGCTAACTGAGGAT','TGGCTAACCCAGGGT','TGGCTAACTCAGGTG','TGGCTAACCCTGGTA','TGGCTAACTTGGGAC','TGGCTAACGCAGGTC','TGGCTAACATCGGTG','TTAATTAAGGAAAAGATCT','CAGAAGAGGATATTAATA','TTGATAAGGAATTGTA','TCAGAGGAGATAATTTA','TGACACGTTGAGCGGTATGA','ACAGATAACAGGAGTAAGTA','TAAAGGGAGAAAAAT','GAGTCTTGAGGTAACTAT','TCAGGAATATTAAAAACGCT','ATTTGAAGGAAAATATT','CAAAAACATACTGCAGGAAT','TGCCATTGCAAAGGAGAAGACT','AAGGGGGAATTCAAAT','AAGGGGTGCAGAAT','AGGTGGAATCACAG','ATAGATAAAAATGGTAACAAT','GGGATATAGCCTGAGGGGCCTGTA','CGGCAATAACAGAGGCGATTT','ATTAAAGAGGAGAAATA','TCACACAGGAAAGTA','AAAGGAGGTGT','AGAGGTGGTGT','AGGAGG','GAGG','TAAAGGAGGAA','AAAGGTGGTGAA','AGGAAACAGAACC','ATATTAAGAGGAGGAG','AGAGAACAAGGAGGGG','GATTGGGATAAATAAT','ATCAACCGGGGTACAT','TTTGGAGATTTTCAAC','AAAAAAGGTAATTCAA','CATAAGGTAATTCACA','ATAAGGAGTCTTAATC','GTTCCGGCTAAGTAAC','TAATGGAAACTTCCTC','TCGCTGGGGGTCAAAG','ATTTGAGGGGGATTCA','AATTTAGGTCAGAAG','AATCAATAGGAGAAATCAAT','TTAAAGAGGAGAAATACTAG']\n",
    "ribosomalbindingsites2=[\"[A-Z]{3}TC[A-Z]{3}TCTTT\",\"TCTCCTCTTTCTTACCCAAAAGCAAGAGGTGATTCTAGTT\",\"TCTCCTCTTTAATACCCAAATCCAAGAGGTGATTCTAGTT\",\"TTTCTCCTCTTTAAT\",'GGTTTCCTGTGTGAA','CACAC','CTAGACACAC','CATCGAACCGGTTTCCTGTGTGA','CATCGAGGCCTTTCCTGTGTGA','CATCCGGCCGTCCTGTGTGA','CTTGACACACGTGAGA','ACACACGTGAGA','AGGGATG','AGGGATGTGA','GGAGGGATGTGA','CTTGTGTTACCTCGTGCAGT','TACTCTTCATCCTCCTCGTA','TACTCCCTCATAAGTAAAGT','TTAGAAGTCTCCGTCTTCGT','TGGTTTACCTCCTGAGGGTT','TAAATGTGTCTCCATGTCTT','CTTGTGTTACCTCGTGCAGT','CAGCTCTTCACATTACTCTTCATCCTCCTC','GATGACCTCCT','GAGACAACTGAAAAAAGAGCTTGC','TCATGCTGGTGATTTTAACTATCAG','TTGTTTTCCTCTTGTTTTTA','AAAGAGGAGA','TTCGCTGCTTCTCCGT','CTTGTCCCAACCTC','TCGGGCTGTCCTTATAAACGGATACATTTA','AGCCGCTCCCACTTAAGAG','AATTTCTCCGGTAGAG','GTAGTCGCTCTTTAACGATGAG','GAAAGGCTCTCCAGGTACAGGCTGAG','CCTCCTTGAG','CCTCTC','TTGTAATCCTCCT','TTTCTCCTCTTT','CTTTCCTGTGTGA','CCTCTTCC','ATTAAAGAGGAGAAA','CGAGGCCTTTCCTGTGTGA','GCTTAATTTCTCCTCTTTAAT','TTTTCAGAAACGAAAATTTTTCAGAAACGA','ACCCTATGTTAGCCA','ATCCTCAGTTAGCCA','ACCCTGGGTTAGCCA','CACCTGAGTTAGCCA','TACCAGGGTTAGCCA','GTCCCAAGTTAGCCA','GACCTGCGTTAGCCA','CACCGATGTTAGCCA','AGATCTTTTCCTTAATTAA','TATTAATATCCTCTTCTG','TACAATTCCTTATCAA','TAAATTATCTCCTCTGA','TCATACCGCTCAACGTGTCA','TACTTACTCCTGTTATCTGT','ATTTTTCTCCCTTTA','ATAGTTACCTCAAGACTC','AGCGTTTTTAATATTCCTGA','AATATTTTCCTTCAAAT','ATTCCTGCAGTATGTTTTTG','AGTCTTCTCCTTTGCAATGGCA','ATTTGAATTCCCCCTT','ATTCTGCACCCCTT','CTGTGATTCCACCT','ATTGTTACCATTTTTATCTAT','TACAGGCCCCTCAGGCTATATCCC','AAATCGCCTCTGTTATTGCCG','TATTTCTCCTCTTTAAT','TACTTTCCTGTGTGA','ACACCTCCTTT','ACACCACCTCT','CCTCCT','CCTC','TTCCTCCTTTA','TTCACCACCTTT','GGTTCTGTTTCCT','CTCCTCCTCTTAATAT','CCCCTCCTTGTTCTCT','ATTATTTATCCCAATC','ATGTACCCCGGTTGAT','GTTGAAAATCTCCAAA','TTGAATTACCTTTTTT','TGTGAATTACCTTATG','GATTAAGACTCCTTAT','GTTACTTAGCCGGAAC','GAGGAAGTTTCCATTA','CTTTGACCCCCAGCGA','TGAATCCCCCTCAAAT','CTTCTGACCTAAATT','ATTGATTTCTCCTATTGATT','CTAGTATTTCTCCTCTTTAA']\n",
    "\n",
    "#create different shine-dalgarno sites:\n",
    "dalgarno1=[]\n",
    "dalgarno2=[]\n",
    "\n",
    "#also look for common cores\n",
    "commoncore=[\"TCCTGGTACATCTGGTACTGA\",\"GCCTGGTACCTCTGGTACTGA\",\"GCCTGGTACATCTGGTACTGA\",\"GCCTGGTACCACTGGTACTAA\"]\n",
    "commoncorerev=[\"TCAGTACCAGATGTACCAGGA\",\"TCAGTACCAGAGGTACCAGGC\",\"TCAGTACCAGATGTACCAGGC\",\"TTAGTACCAGTGGTACCAGGC\"]\n",
    "#create reading frame patterns\n",
    "for i in range(0,15):\n",
    "    dalgarno1=dalgarno1+[\"AGGAGG[A-Z]{\"+str(i)+\"}\"]\n",
    "    dalgarno2=dalgarno2+[\"[A-Z]{\"+str(i)+\"}CCTCCT\"]\n",
    "ribosomalbindingsites1=ribosomalbindingsites1+dalgarno1\n",
    "ribosomalbindingsites2=ribosomalbindingsites2+dalgarno2\n",
    "for i in range(minlengthreadingframe,maxlengthreadingframe):\n",
    "    i=i*3\n",
    "    for end in stopcodon1:\n",
    "        for n in range (4,8):\n",
    "            for start in startcodon1:\n",
    "                for rbs in ribosomalbindingsites1:\n",
    "\n",
    "                    listreadingframe=listreadingframe+[\"(?<=\"+str(rbs)+\"[A-Z]{\"+str(n)+\"}\"+\")\"+start+\"[A-Z]{\"+str(i)+\"}\"+str(end)]\n",
    "for i in range(minlengthreadingframe,maxlengthreadingframe):\n",
    "    i=i*3\n",
    "    for end in stopcodon2:\n",
    "        for n in range (4,8):\n",
    "            for start in startcodon2:\n",
    "                for rbs in ribosomalbindingsites2:\n",
    "                    newframe=end+\"[A-Z]{\"+str(i)+\"}\"+start+\"(?=[A-Z]{\"+str(n)+\"}\"+str(rbs)+\")\"\n",
    "                    listreadingframe=listreadingframe+ [newframe]\n",
    "for i in range(minlengthreadingframe-7,maxlengthreadingframe):\n",
    "    i=i*3\n",
    "    for end in commoncore:\n",
    "        for start in startcodon1:\n",
    "            listreadingframe=listreadingframe+[start+\"[A-Z]{\"+str(i)+\"}\"+str(end)]\n",
    "for i in range(minlengthreadingframe-7,maxlengthreadingframe):\n",
    "    i=i*3\n",
    "    for end in commoncorerev:\n",
    "        for start in startcodon2:\n",
    "            newframe=end+\"[A-Z]{\"+str(i)+\"}\"+start\n",
    "            listreadingframe=listreadingframe+ [newframe]\n",
    "#safe file\n",
    "with open(filenamereadingframes, 'w') as f:\n",
    "    for s in listreadingframe:\n",
    "        f.write(str(s) + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "surprised-wealth",
   "metadata": {},
   "source": [
    "# Find reading frames in genomes next to P450s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deadly-supervisor",
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "import re\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio import Entrez\n",
    "\n",
    "#tableofproteins.to_csv(filename_output, index=False)\n",
    "\n",
    "front=2000\n",
    "back=2000\n",
    "\n",
    "stopcodon1=('TGA', 'TAA', 'TAG')\n",
    "startcodon1=('ATG',\"GTG\")\n",
    "stopcodon2=(\"TCA\",\"TTA\",\"CTA\")\n",
    "startcodon2=(\"CAT\",\"CAC\")\n",
    "\n",
    "#insert length of protein in aa\n",
    "minlengthreadingframe=20\n",
    "maxlengthreadingframe=40\n",
    "\n",
    "filename_input=\"Output/sirinsp450permutationsof4/p450ncbiallpreoteins.fastalistpositive.fasta\"\n",
    "filename_output=filename_input+\"infonew.csv\"\n",
    "tableofproteins=pd.DataFrame(columns=['Protein Reference','NCBI Reference','Species',\"Definition\", \"gb_whole_genome_accession\",\"gb_whole_genome_position\",\"seqencednafragment\",\"openreadingframes\",\"coreproteins\"])\n",
    "Entrez.email = \"friederike@biermann-erfurt.de\"\n",
    "filenamereadingframes=\"Output/sirinsp450permutationsof4/readingframes.txt\"\n",
    "listfasta=[]\n",
    "liststrains=[]\n",
    "\n",
    "def find_between( s, first, last ):\n",
    "    try:\n",
    "        start = s.index( first ) + len( first )\n",
    "        end = s.index( last, start )\n",
    "        return s[start:end]\n",
    "    except ValueError:\n",
    "        return \"\"\n",
    "def find_infrontof( s, last ):\n",
    "    try:\n",
    "        start = 0\n",
    "        end = s.index( last, start )\n",
    "        return s[start:end]\n",
    "    except ValueError:\n",
    "        return \"\"\n",
    "def find_after( s, first ):\n",
    "    try:\n",
    "        start =s.index( first ) + len( first )\n",
    "        end = len(s)\n",
    "        return s[start:end]\n",
    "    except ValueError:\n",
    "        return \"\"\n",
    "    \n",
    "listreadingframe=[]\n",
    "with open(filenamereadingframes, 'r') as f:\n",
    "    listreadingframe = [line.rstrip('\\n') for line in f]\n",
    "    \n",
    "#searching for reading frames in fastas\n",
    "for record in SeqIO.parse(filename_input, \"fasta\"):\n",
    "    try:\n",
    "        rid=record.id\n",
    "        print (rid)\n",
    "        if \"ref\" in rid:\n",
    "            #find id in fasta header\n",
    "            gi = find_between(record.id,\"gi|\",\"|\")\n",
    "            ref= find_between(record.id,\"ref|\",\"|\")\n",
    "            # find all nucleotides this reference protein is part of\n",
    "            handle = Entrez.elink(dbFrom = \"protein\", db = \"nucleotide\",id=gi)\n",
    "            record = str(Entrez.read(handle))\n",
    "            handle.close()\n",
    "            listgenes=re.findall(\"(?<=Id': ')[0-9]+\",record)\n",
    "            listgenes=listgenes[:int(len(listgenes)/2)]\n",
    "            for name1 in listgenes:\n",
    "                # get sequence of every genome\n",
    "                dnahandle1 = Entrez.efetch(db=\"nucleotide\", id=name1, rettype='gbwithparts',retmode=\"xml\")\n",
    "                record = Entrez.read(dnahandle1)\n",
    "                dnahandle1.close()\n",
    "                definition=str(record[0][\"GBSeq_definition\"])\n",
    "                species = record[0][\"GBSeq_organism\"]\n",
    "                for match in re.finditer(ref, str(record)):\n",
    "                    #find p450 in sequence/ in genbank file and the proteins before and after\n",
    "                    proteinspan1=match.span()[0]\n",
    "                   # print (proteinspan1)\n",
    "                    proteinspan2=int((proteinspan1))-1000\n",
    "                    proteinspan3=int((proteinspan1))-3000\n",
    "                    proteinspan4=int((match.span()[1]))+1000\n",
    "                    startdna=re.search(\"(?<='GBInterval_from': ')[0-9]+\", str(record)[proteinspan2:proteinspan1])\n",
    "                    enddna=re.search(\"(?<='GBInterval_to': ')[0-9]+\", str(record)[proteinspan2:proteinspan1])                    \n",
    "                    startdna2=re.search(\"(?<='GBInterval_from': ')[0-9]+\", str(record)[match.span()[1]:proteinspan4])                 \n",
    "                    enddna2=re.search(\"(?<='GBInterval_to': ')[0-9]+\", str(record)[match.span()[1]:proteinspan4])                    \n",
    "                    startdna3=re.search(\"(?<='GBInterval_from': ')[0-9]+\", str(record)[proteinspan3:proteinspan2])\n",
    "                    enddna3=re.search(\"(?<='GBInterval_to': ')[0-9]+\", str(record)[proteinspan3:proteinspan2])\n",
    "                    p450count=0\n",
    "                    listp450=[]\n",
    "                    \n",
    "                    #find exact p450\n",
    "                    for match in re.finditer(\"P450\", str(record)):\n",
    "                        p450count=p450count+1\n",
    "                        proteinspan1=match.span()[0]+500\n",
    "                        proteinspan2=int((proteinspan1))-1000\n",
    "                        startp450=re.search(\"(?<='GBInterval_from': ')[0-9]+\", str(record)[proteinspan2:proteinspan1])  \n",
    "                        endp450=re.search(\"(?<='GBInterval_to': ')[0-9]+\", str(record)[proteinspan2:proteinspan1])                        \n",
    "                        startp450=startp450.group(0)\n",
    "                        endp450=endp450.group(0)\n",
    "                        string= str(startp450)+\"-\"+str(endp450)\n",
    "                        listp450.append(string)\n",
    "                    startdna=startdna.group(0)\n",
    "                    enddna=enddna.group(0)\n",
    "                    startdna2=startdna2.group(0)\n",
    "                    enddna2=enddna2.group(0)\n",
    "                    startdna3=startdna3.group(0)\n",
    "                    enddna3=enddna3.group(0)\n",
    "                if len(enddna)>0 and len(startdna)>0:\n",
    "                    framebegin=int(startdna)-front\n",
    "                    frameend=int(enddna)+back\n",
    "                    #get sequence of whole genomic background beore/after protein\n",
    "                    dnahandle = Entrez.efetch(db=\"nucleotide\", id=name1, rettype=\"fasta\",retmode=\"txt\")\n",
    "                    for seq_record in SeqIO.parse(dnahandle, \"fasta\"):\n",
    "                        complete_sequence=seq_record.seq\n",
    "                        chosensequence=str(complete_sequence[framebegin:frameend])\n",
    "                    dnahandle.close()\n",
    "                    listopenframe=\"\"\n",
    "                    listcore=\"\"\n",
    "                    for frame in listreadingframe:\n",
    "                        #actual corefinder- searches for open reading frames near protein\n",
    "                        for match in re.finditer(frame,chosensequence):\n",
    "                           # print(len(match.group()))\n",
    "                            begin=match.span()[0]+framebegin\n",
    "                            end=match.span()[1]+framebegin\n",
    "                            # make sure core is not in things annotated in genbank file\n",
    "                            if int(startdna)<begin<int(enddna)or int(startdna)<end<int(enddna)or int(enddna)<begin<int(startdna)or int(enddna)<end<int(startdna) :\n",
    "                                   pass\n",
    "                            if int(startdna2)<begin<int(enddna2)or int(startdna2)<end<int(enddna2)or int(enddna2)<begin<int(startdna2)or int(enddna2)<end<int(startdna2) :\n",
    "                                   pass\n",
    "                            if int(startdna3)<begin<int(enddna3)or int(startdna3)<end<int(enddna3)or int(enddna3)<begin<int(startdna3)or int(enddna3)<end<int(startdna3) :\n",
    "                                   pass\n",
    "                            else:                               \n",
    "                                if match.group()[:3] in startcodon1:\n",
    "                                    # if forward strand\n",
    "                                    coredna=Seq(match.group()[-21:])\n",
    "                                    aa=Seq(match.group()).translate()\n",
    "                                    coreaa=coredna.translate()\n",
    "                                     #tests if stop codon included\n",
    "                                    if \"*\" in aa[:-1]:\n",
    "                                         pass\n",
    "                                    if \"*\" not in aa[:-1] and \"*\" not in coreaa[:-1]:\n",
    "                                        print (\"+\",coreaa)\n",
    "                                        if str(end) and str(begin) not in listcore:\n",
    "                                            listopenframe=listopenframe+\", \"+match.group()+\", \"+str(begin)+\", \"+str(end)+\"(+)\"\n",
    "                                            listcore=listcore+\", \"+str(coreaa)+\", \"+str(begin)+\", \"+str(end)+\"(+)\"\n",
    "                                if match.group()[:3] in stopcodon2:\n",
    "                                    #if reverse strand\n",
    "                                    reversedna=Seq(match.group()).back_transcribe().reverse_complement()\n",
    "                                    reversedna=str(reversedna)\n",
    "                                    aa=Seq(reversedna).translate()\n",
    "                                    coredna=reversedna[-21:]\n",
    "                                    coreaa=Seq(coredna).translate()\n",
    "                                    #tests if stop codon included\n",
    "                                    if \"*\" in aa[:-1]:\n",
    "                                         pass\n",
    "                                    if \"*\" not in aa[:-1] and \"*\" not in coreaa[:-1]:\n",
    "                                        if str(end) and str(begin) not in listcore:\n",
    "                                            print (\"-\",coreaa)\n",
    "                                            listopenframe=listopenframe+\", \"+str(reversedna)+\", \"+str(begin)+\", \"+str(end)+\"(-)\"\n",
    "                                            listcore=listcore+\", \"+str(coreaa)+\", \"+str(begin)+\", \"+str(end)+\"(-)\"\n",
    "                    position4=str(startdna)+\"..\"+str(enddna)\n",
    "                    # add new row with information to table\n",
    "                    new_row={'Protein Reference':ref,'NCBI Reference':name1,'Species':species,\"Definition\":definition,\"gb_whole_genome_accession\":name1,\"gb_whole_genome_position\":position4,\"seqencednafragment\":chosensequence,\"openreadingframes\":str(listopenframe),\"coreproteins\":str(listcore)}\n",
    "                    \n",
    "                    tableofproteins = tableofproteins.append(new_row, ignore_index=True)\n",
    "        \n",
    "        if \"gb\" in rid:\n",
    "            # same thing but not with refernce protein but with exact protein\n",
    "            gi = find_between(record.id,\"gi|\",\"|\")\n",
    "            ref= find_between(record.id,\"gb|\",\"|\")\n",
    "            handle = Entrez.efetch(db=\"protein\", id=gi, rettype=\"gb\", retmode=\"xml\")\n",
    "            records = Entrez.parse(handle)            \n",
    "            for record in records:\n",
    "                definition=record[\"GBSeq_definition\"]\n",
    "                name = record[\"GBSeq_organism\"]\n",
    "                name1=find_after(name,\"n \")\n",
    "                position = str(record['GBSeq_feature-table'])\n",
    "                position2=find_between(position, \"locus_tag\",\"transl_table\")\n",
    "                position3=find_between(position2, \"{'GBQualifier_name': 'coded_by', 'GBQualifier_value': '\",\"'}, {'GBQualifier_name': 'note'\") \n",
    "                position4=find_after(position3,\":\")[0:25]\n",
    "                species = record[\"GBSeq_organism\"]\n",
    "                #get dna of strain\n",
    "                startdna=find_infrontof( position4, \".\")\n",
    "                enddna=find_after(position4, \"..\")\n",
    "                startdna= re.findall(r'\\d', startdna)\n",
    "                startdna=\"\".join(startdna)\n",
    "                enddna= re.findall(r'\\d', enddna)\n",
    "                enddna=\"\".join(enddna)\n",
    "                if len(enddna)>0 and len(startdna)>0:\n",
    "                    framebegin=int(startdna)-front\n",
    "                    frameend=int(enddna)+back\n",
    "                    dnahandle = Entrez.efetch(db=\"nucleotide\", id=name1, rettype=\"fasta\",retmode=\"txt\")\n",
    "                    chosensequence=\"\"\n",
    "                    for seq_record in SeqIO.parse(dnahandle, \"fasta\"):\n",
    "                        complete_sequence=seq_record.seq\n",
    "                        chosensequence=str(complete_sequence[framebegin:frameend])\n",
    "                    dnahandle.close()\n",
    "                    listopenframe=\"\"\n",
    "                    listcore=\"\"\n",
    "                    for frame in listreadingframe:\n",
    "                        for match in re.finditer(frame,chosensequence):            \n",
    "                            begin=match.span()[0]+framebegin\n",
    "                            end=match.span()[1]+framebegin\n",
    "                            if int(startdna)<begin<int(enddna)or int(startdna)<end<int(enddna)or int(enddna)<begin<int(startdna)or int(enddna)<end<int(startdna) :\n",
    "                                   pass\n",
    "                            else:\n",
    "                                if match.group()[:3] ==\"ATG\":\n",
    "                                    aa=Seq(match.group()).translate()\n",
    "                                    coreaa=aa[-7:]\n",
    "                                    if \"*\" in aa[:-1]:\n",
    "                                         pass\n",
    "                                    if \"*\" not in aa[:-1] and \"*\" not in coreaa[:-1]:\n",
    "                                        if str(end) and str(begin) not in listcore:\n",
    "                                            listopenframe=listopenframe+\", \"+match.group()+\", \"+str(begin)+\", \"+str(end)+\"(+)\"\n",
    "                                            listcore=listcore+\", \"+str(coreaa)+\", \"+str(begin)+\", \"+str(end)+\"(+)\"\n",
    "                                if match.group()[:3] in stopcodon2:\n",
    "                                    reversedna=Seq(match.group()).back_transcribe().reverse_complement()\n",
    "                                    reversedna=str(reversedna)\n",
    "                                    aa=Seq(reversedna).translate()\n",
    "                                    coreaa=aa[-7:]\n",
    "                                    if \"*\" in aa[:-1]:\n",
    "                                         pass\n",
    "                                    if \"*\" not in aa[:-1] and \"*\" not in coreaa[:-1]:\n",
    "                                        if str(end) and str(begin) not in listcore:\n",
    "                                            print (coreaa)\n",
    "                                            listopenframe=listopenframe+\", \"+str(reversedna)+\", \"+str(begin)+\", \"+str(end)+\"(-)\"\n",
    "                                            listcore=listcore+\", \"+str(coreaa)+\", \"+str(begin)+\", \"+str(end)+\"(-)\"\n",
    "                    position4=str(startdna)+':'+str(enddna)\n",
    "                    new_row={'Protein Reference':ref,'NCBI Reference':name1,'Species':species,\"Definition\":definition,\"gb_whole_genome_accession\":name1,\"gb_whole_genome_position\":position4,\"seqencednafragment\":chosensequence,\"openreadingframes\":str(listopenframe),\"coreproteins\":str(listcore)}\n",
    "                    tableofproteins = tableofproteins.append(new_row, ignore_index=True)\n",
    "                    handle.close()\n",
    "    except: pass\n",
    "tableofproteins.to_csv(filename_output, index=False) "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
