{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract ferredoxins and P450 from clusters annotated by antismash 5.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "import re\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import Entrez\n",
    "import os\n",
    "import glob\n",
    "\n",
    "#fill in correct filenames for input/output files\n",
    "path = 'Output/'\n",
    "filenameout=\"tablep450ferredox.csv\"\n",
    "output=\"ferredox.fasta\"\n",
    "output1=\"p450ferre.fasta\"\n",
    "\n",
    "#creates empty data frame\n",
    "tableofproteins=pd.DataFrame(columns=[\"accession\",'p450', 'ferredoxin'])\n",
    "listp450=[]\n",
    "listferredox=[]\n",
    "\n",
    "#process antismash file\n",
    "for gb_file in glob.glob(os.path.join(path, '*.gbk')):\n",
    "    \n",
    "        gb_record = SeqIO.read(open(gb_file,\"r\"), \"genbank\")\n",
    "        p450=[]\n",
    "        ferredox=[]\n",
    "        for feature in gb_record.features:\n",
    "            if feature.type==\"CDS\":\n",
    "                try:\n",
    "                    if \"ferredoxin\" in feature.qualifiers['product'][0]:\n",
    "                        pid=feature.qualifiers['protein_id'][0]\n",
    "                        transl=feature.qualifiers['translation'][0]\n",
    "                        listferredox.append(SeqRecord(Seq(transl),id=pid))\n",
    "                        ferredox.append(pid)\n",
    "                    if \"P450\" in feature.qualifiers['product'][0]:\n",
    "                        pid=feature.qualifiers['protein_id'][0]\n",
    "                        transl=feature.qualifiers['translation'][0]\n",
    "                        listp450.append(SeqRecord(Seq(transl),id=pid))\n",
    "                        p450.append(pid)\n",
    "                except: pass\n",
    "        new_row={\"accession\":gb_record.name,\"p450\":str(p450),\"ferredoxin\":str(ferredox)}\n",
    "        tableofproteins = tableofproteins.append(new_row, ignore_index=True)\n",
    "                    \n",
    "#safe files\n",
    "SeqIO.write(listp450, output1, 'fasta')\n",
    "SeqIO.write(listferredox, output, 'fasta')\n",
    "tableofproteins.to_csv(filenameout, index=False) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get ferredoxins associated with P450s from NCBI files (fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio import Entrez\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "filename_input=\"Fasta/ferredoxinncbi.fasta\"\n",
    "filename_output=filename_input+\"info.csv\"\n",
    "outputferre=filename_input+\"ferre.fasta\"\n",
    "outputp450=filename_input+\"p450.fasta\"\n",
    "tableofproteins=pd.DataFrame(columns=['Protein Reference','Acessiongenome','Species',\"Definition\",\"proteins\",\"blasts\",\"products\"])\n",
    "Entrez.email = \"friederike@biermann-erfurt.de\"\n",
    "listfasta=[]\n",
    "liststrains=[]\n",
    "\n",
    "def find_between( s, first, last ):\n",
    "    try:\n",
    "        start = s.index( first ) + len( first )\n",
    "        end = s.index( last, start )\n",
    "        return s[start:end]\n",
    "    except ValueError:\n",
    "        return \"\"\n",
    "def find_infrontof( s, last ):\n",
    "    try:\n",
    "        start = 0\n",
    "        end = s.index( last, start )\n",
    "        return s[start:end]\n",
    "    except ValueError:\n",
    "        return \"\"\n",
    "def find_after( s, first ):\n",
    "    try:\n",
    "        start =s.index( first ) + len( first )\n",
    "        end = len(s)\n",
    "        return s[start:end]\n",
    "    except ValueError:\n",
    "        return \"\"\n",
    "\n",
    "listp450=[]\n",
    "listferredoxins=[]\n",
    "countsafe=0\n",
    "\n",
    "#actually finding in fastas\n",
    "for record in SeqIO.parse(filename_input, \"fasta\"):\n",
    "    countsafe=countsafe+1\n",
    "    try:\n",
    "            if \"ref\" in record.id:\n",
    "                gi=find_between(record.id, \"ref|\",\"|\")\n",
    "            if \"gb\" in record.id:\n",
    "                gi=find_between(record.id, \"gb|\",\"|\") \n",
    "            if \"sp\" in record.id:\n",
    "                pi=find_between(record.id, \"sp|\",\"|\") \n",
    "                handle = Entrez.elink(dbFrom = \"protein\", db = \"protein\",id=pi)\n",
    "                recordtest = str(Entrez.read(handle))\n",
    "                handle.close()\n",
    "                for match in re.finditer(\"(?<=Id': ')[0-9]+\", recordtest):\n",
    "                            gi=match.group()\n",
    "            ref=str(record.seq)\n",
    "            # fetch all genomes associated with protein\n",
    "            handle = Entrez.elink(dbFrom = \"protein\", db = \"nucleotide\",id=gi)\n",
    "            record3 = str(Entrez.read(handle))\n",
    "            handle.close()\n",
    "            listgenes=re.findall(\"(?<=Id': ')[0-9]+\",record3)\n",
    "            listgenes=listgenes[:int(len(listgenes)/2)]\n",
    "            for name1 in listgenes[:2]:\n",
    "                dnahandle1 = Entrez.efetch(db=\"nucleotide\", id=name1, rettype='gbwithparts',retmode=\"xml\")\n",
    "                record4 = Entrez.read(dnahandle1)\n",
    "                dnahandle1.close()\n",
    "                definition=str(record4[0][\"GBSeq_definition\"])\n",
    "                species = record4[0][\"GBSeq_organism\"]\n",
    "                listproteins=[]\n",
    "                listsequences=[]\n",
    "                listproducts=[]\n",
    "                # find protein of interrest in genome and search neighbouring genes for \"P450\"\n",
    "                for match in re.finditer(ref, str(record4)):\n",
    "                    proteinspan1=match.span()[0]\n",
    "                    proteinspan2=int((proteinspan1))-25000\n",
    "                    proteinspan3=int((proteinspan1))+27000\n",
    "                    intervalrecord=str(record4)[proteinspan2:proteinspan3]                   \n",
    "                    if \"P450\" in intervalrecord or \"p450\" in intervalrecord:\n",
    "                        for match in re.finditer(\"(?<={'GBQualifier_name': 'product', 'GBQualifier_value': ')...................\", intervalrecord):\n",
    "                            if \"P450\" in match.group() or \"p450\" in match.group():\n",
    "                                interval2= intervalrecord[match.span()[0]-1000:match.span()[0]+1000]\n",
    "                                listproducts.append(match.group())\n",
    "                                for match in re.finditer(\"(?<={'GBQualifier_name': 'translation', 'GBQualifier_value': ')\\w+\", interval2):\n",
    "                                    listsequences.append(match.group())\n",
    "                                    translation=match.group()\n",
    "                                for match in re.finditer(\"(?<={'GBQualifier_name': 'locus_tag', 'GBQualifier_value': ')\\w+\", interval2):\n",
    "                                    listproteins.append(match.group())\n",
    "                                    locus=match.group()\n",
    "                                listp450.append(SeqRecord(Seq(translation),id=locus))\n",
    "                                listferredoxins.append(record)\n",
    "                        res = [] \n",
    "                        [res.append(x) for x in listproteins if x not in res]\n",
    "                        \n",
    "                        #safe results\n",
    "                        new_row={'Protein Reference':gi,'Acessiongenome':name1,'Species':species,\"Definition\":definition,\"proteins\":res,\"blasts\":listsequences,\"products\":listproducts}\n",
    "                        tableofproteins = tableofproteins.append(new_row, ignore_index=True)\n",
    "                        \n",
    "                        #safe every 100 entries\n",
    "                        if countsafe% 100 == 0:\n",
    "                            tableofproteins.to_csv(filename_output, index=False) \n",
    "                            SeqIO.write(listferredoxins, outputferre, 'fasta')\n",
    "                            SeqIO.write(listp450, outputp450, 'fasta')\n",
    "    except: pass\n",
    "      \n",
    "#final saving of files\n",
    "tableofproteins.to_csv(filename_output, index=False) \n",
    "SeqIO.write(listferredoxins, outputferre, 'fasta')\n",
    "SeqIO.write(listp450, outputp450, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get Ferredoxins, that are not associated with p450s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WP_207629503.1\n",
      "WP_207628008.1\n",
      "1821604050\n",
      "8112195\n",
      "a\n",
      "zu nah\n",
      "[ExactPosition(252359), ExactPosition(7857552), ExactPosition(3048145), ExactPosition(722390), ExactPosition(7833625), ExactPosition(7765659), ExactPosition(7256412), ExactPosition(253596), ExactPosition(7808414), ExactPosition(1139423), ExactPosition(7809634), ExactPosition(481188), ExactPosition(5427581), ExactPosition(5426343), ExactPosition(5158187), ExactPosition(4151100), ExactPosition(6829565)] [ExactPosition(4397442), ExactPosition(1714693), ExactPosition(6514483), ExactPosition(7556822), ExactPosition(6806782)]\n",
      "WP_207623627.1\n",
      "WP_207619503.1\n",
      "1823026339\n",
      "81237\n",
      "WP_207598691.1\n",
      "2007941048\n",
      "357176\n",
      "WP_207598691.1\n",
      "2007941048\n",
      "357176\n",
      "WP_207598691.1\n",
      "2007941048\n",
      "357176\n",
      "WP_207576506.1\n",
      "1184616905\n",
      "143188\n",
      "WP_207566703.1\n",
      "WP_207565566.1\n",
      "2007832697\n",
      "466817\n",
      "WP_207564100.1\n",
      "2007832696\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-5cc59b6c84e5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     91\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 92\u001b[0;31m                 \u001b[0mrecord4\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdnahandle1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"genbank\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     93\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     94\u001b[0m                 \u001b[0;31m#print (record4)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/site-packages/Bio/SeqIO/__init__.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    654\u001b[0m     \u001b[0miterator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malphabet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    655\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 656\u001b[0;31m         \u001b[0mrecord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterator\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    657\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    658\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"No records found in handle\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/site-packages/Bio/SeqIO/Interfaces.py\u001b[0m in \u001b[0;36m__next__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     71\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__next__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     72\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 73\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecords\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     74\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     75\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshould_close_stream\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/site-packages/Bio/GenBank/Scanner.py\u001b[0m in \u001b[0;36mparse_records\u001b[0;34m(self, handle, do_features)\u001b[0m\n\u001b[1;32m    514\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mas_handle\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    515\u001b[0m             \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 516\u001b[0;31m                 \u001b[0mrecord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdo_features\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    517\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mrecord\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    518\u001b[0m                     \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/site-packages/Bio/GenBank/Scanner.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(self, handle, do_features)\u001b[0m\n\u001b[1;32m    497\u001b[0m         )\n\u001b[1;32m    498\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 499\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfeed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconsumer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdo_features\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    500\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mconsumer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    501\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/site-packages/Bio/GenBank/Scanner.py\u001b[0m in \u001b[0;36mfeed\u001b[0;34m(self, handle, consumer, do_features)\u001b[0m\n\u001b[1;32m    468\u001b[0m         \u001b[0;31m# Features (common to both EMBL and GenBank):\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    469\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mdo_features\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 470\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_feed_feature_table\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconsumer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse_features\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mskip\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    471\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    472\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse_features\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mskip\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# ignore the data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/site-packages/Bio/GenBank/Scanner.py\u001b[0m in \u001b[0;36mparse_features\u001b[0;34m(self, skip)\u001b[0m\n\u001b[1;32m    227\u001b[0m                     \u001b[0;31m# white space (e.g. out of spec files with too much indentation)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    228\u001b[0m                     \u001b[0mfeature_lines\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFEATURE_QUALIFIER_INDENT\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 229\u001b[0;31m                     \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    230\u001b[0m                 \u001b[0mfeatures\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse_feature\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfeature_key\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeature_lines\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    231\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36mread1\u001b[0;34m(self, n)\u001b[0m\n\u001b[1;32m    630\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0;34mb\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    631\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchunked\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 632\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read1_chunked\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    633\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlength\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m0\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlength\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    634\u001b[0m             \u001b[0mn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlength\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36m_read1_chunked\u001b[0;34m(self, n)\u001b[0m\n\u001b[1;32m    667\u001b[0m         \u001b[0;31m# Strictly speaking, _get_chunk_left() may cause more than one read,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    668\u001b[0m         \u001b[0;31m# but that is ok, since that is to satisfy the chunked protocol.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 669\u001b[0;31m         \u001b[0mchunk_left\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_chunk_left\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    670\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mchunk_left\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    671\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0;34mb''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36m_get_chunk_left\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    551\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mchunk_left\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    552\u001b[0m                 \u001b[0;31m# We are at the end of chunk, discard chunk end\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 553\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_safe_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# toss the CRLF at the end of the chunk\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    554\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    555\u001b[0m                 \u001b[0mchunk_left\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_next_chunk_size\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36m_safe_read\u001b[0;34m(self, amt)\u001b[0m\n\u001b[1;32m    610\u001b[0m         \u001b[0mIncompleteRead\u001b[0m \u001b[0mexception\u001b[0m \u001b[0mcan\u001b[0m \u001b[0mbe\u001b[0m \u001b[0mused\u001b[0m \u001b[0mto\u001b[0m \u001b[0mdetect\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mproblem\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    611\u001b[0m         \"\"\"\n\u001b[0;32m--> 612\u001b[0;31m         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mamt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    613\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mamt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    614\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mIncompleteRead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mamt\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/socket.py\u001b[0m in \u001b[0;36mreadinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    667\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    668\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 669\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    670\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    671\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_timeout_occurred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/ssl.py\u001b[0m in \u001b[0;36mrecv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1239\u001b[0m                   \u001b[0;34m\"non-zero flags not allowed in calls to recv_into() on %s\"\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1240\u001b[0m                   self.__class__)\n\u001b[0;32m-> 1241\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbytes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1242\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1243\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbuffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnbytes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/Diplomarbeit/lib/python3.8/ssl.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1097\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1098\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mbuffer\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1099\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1100\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1101\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio import Entrez\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "# fill in correct input/output file names\n",
    "filename_input=\"Fasta/ferredoxinncbi.fasta\"\n",
    "filename_output=filename_input+\"infoeinzel2.csv\"\n",
    "outputferre=filename_input+\"ferreeinzel2.fasta\"\n",
    "outputp450=filename_input+\"p450einzel2.fasta\"\n",
    "tableofproteins=pd.DataFrame(columns=['Protein Reference','Acessiongenome',\"p450 product\",\"p450 accessions\",\"p450 loci\",\"p450 translations\",\"ferredoxin product\",\"ferredoxin accession\",\"ferredoxin loci\",\"ferredoxin translation\"])\n",
    "Entrez.email = \"friederike@biermann-erfurt.de\"\n",
    "listfasta=[]\n",
    "liststrains=[]\n",
    "\n",
    "a=700\n",
    "b=10000\n",
    "def find_between( s, first, last ):\n",
    "    try:\n",
    "        start = s.index( first ) + len( first )\n",
    "        end = s.index( last, start )\n",
    "        return s[start:end]\n",
    "    except ValueError:\n",
    "        return \"\"\n",
    "def find_infrontof( s, last ):\n",
    "    try:\n",
    "        start = 0\n",
    "        end = s.index( last, start )\n",
    "        return s[start:end]\n",
    "    except ValueError:\n",
    "        return \"\"\n",
    "def find_after( s, first ):\n",
    "    try:\n",
    "        start =s.index( first ) + len( first )\n",
    "        end = len(s)\n",
    "        return s[start:end]\n",
    "    except ValueError:\n",
    "        return \"\"\n",
    "def list_diff(my_list1, my_list2):\n",
    "       return (list(set(my_list1) - set(my_list2)))\n",
    "\n",
    "listp450=[]\n",
    "listferredoxins=[]\n",
    "countsafe=0\n",
    "#actually finding in fastas\n",
    "for record in SeqIO.parse(filename_input, \"fasta\"):\n",
    "    \n",
    "    try:\n",
    "            countsafe=countsafe+1\n",
    "            if \"ref\" in record.id:\n",
    "                gi=find_between(record.id, \"ref|\",\"|\")\n",
    "            if \"gb\" in record.id:\n",
    "                gi=find_between(record.id, \"gb|\",\"|\") \n",
    "            if \"sp\" in record.id:\n",
    "                pi=find_between(record.id, \"sp|\",\"|\") \n",
    "                handle = Entrez.elink(dbFrom = \"protein\", db = \"protein\",id=pi)\n",
    "                recordtest = str(Entrez.read(handle))\n",
    "                handle.close()\n",
    "                for match in re.finditer(\"(?<=Id': ')[0-9]+\", recordtest):\n",
    "                            gi=match.group()\n",
    "            ref=str(record.seq)\n",
    "            handle = Entrez.elink(dbFrom = \"protein\", db = \"nucleotide\",id=gi)\n",
    "            record3 = str(Entrez.read(handle))\n",
    "            handle.close()\n",
    "            listgenes=re.findall(\"(?<=Id': ')[0-9]+\",record3)\n",
    "            listgenes=listgenes[:int(len(listgenes)/2)]\n",
    "            for name1 in listgenes[:2]:\n",
    "                dnahandle1 = Bio.Entrez.efetch(db=\"nucleotide\", id=name1, rettype='gbwithparts',retmode=\"text\")\n",
    "                record4 = SeqIO.read(dnahandle1, \"genbank\")\n",
    "                dnahandle1.close()\n",
    "                listp450proteins=[]\n",
    "                listp450sequences=[]\n",
    "                listp450products=[]\n",
    "                listferredoxinproteins=[]\n",
    "                listferredoxinsequences=[]\n",
    "                listferredoxinproducts=[]\n",
    "                listp450loci=[]\n",
    "                listferredoxinloci=[]\n",
    "                intervalrecord=str(record4)\n",
    "                if len(record4.seq)>2000000:\n",
    "                    for feature in record4.features:\n",
    "                        if feature.type==\"CDS\":\n",
    "                            if \"ferredoxin\" in feature.qualifiers['product'][0]:\n",
    "                                try:\n",
    "                                    if len(feature.qualifiers['translation'][0])<150:\n",
    "                                        product=feature.qualifiers['product'][0]\n",
    "                                        pid=feature.qualifiers['protein_id'][0]\n",
    "                                        locus=feature.location\n",
    "                                        transl=feature.qualifiers['translation'][0]\n",
    "                                        listferredoxinsequences.append(transl)\n",
    "                                        listferredoxinproducts.append(product)\n",
    "                                        listferredoxinloci.append(locus)\n",
    "                                        listferredoxinproteins.append(pid)\n",
    "                                except: pass\n",
    "                            if \"P450\" in feature.qualifiers['product'][0]:\n",
    "                                try:\n",
    "                                    #print(feature.qualifiers)\n",
    "                                    product=feature.qualifiers['product'][0]\n",
    "                                    pid=feature.qualifiers['protein_id'][0]\n",
    "                                    transl=feature.qualifiers['translation'][0]\n",
    "                                    locus=feature.location\n",
    "                                    listp450proteins.append(pid)\n",
    "                                    listp450sequences.append(transl)\n",
    "                                    listp450products.append(product)\n",
    "                                    listp450loci.append(locus)\n",
    "                                except: pass\n",
    "                    # compare loci list of ferredoxins/p450, searche for isolated ones\n",
    "                    templistferredoxins=[]\n",
    "                    templistp450=[]\n",
    "                    listp450locin=[]\n",
    "                    for locus in listp450loci:\n",
    "                         listp450locin.append(locus.start)\n",
    "                    listferredoxinlocin=[]\n",
    "                    for locus in listferredoxinloci:\n",
    "                            listferredoxinlocin.append(locus.start)\n",
    "                    if len(listferredoxinloci)>1:\n",
    "                        for ferredoxin in listferredoxinloci:\n",
    "                            c=int(ferredoxin.start)+b\n",
    "                            d=int(ferredoxin.end)-b\n",
    "                            for p450 in listp450loci:\n",
    "                                if c>int(p450.start)>d or c>int(p450.end)>d :\n",
    "                                    print (\"zu nah\")\n",
    "                                    if p450 not in templistp450:\n",
    "                                        templistp450.append(p450.start)\n",
    "                                    if ferredoxin not in templistferredoxins:\n",
    "                                        templistferredoxins.append(ferredoxin.start)\n",
    "                        listnewp450=list_diff(listp450locin,templistp450)\n",
    "                        listnewferredoxin=list_diff(listferredoxinlocin,templistferredoxins)\n",
    "                    else:\n",
    "                        listnewp450=listp450locin\n",
    "                        listnewferredoxin=listferredoxinlocin\n",
    "                    if len(listnewferredoxin)==1 and len(listnewp450)>0:\n",
    "                        countsafe=countsafe+1\n",
    "                        for index, value in enumerate(listferredoxinlocin):\n",
    "                                if value in listnewferredoxin:\n",
    "                                    translationf=listferredoxinsequences[index]\n",
    "                                    locusf=listferredoxinloci[index]\n",
    "                                    IDf=listferredoxinproteins[index]\n",
    "                                    productf=listferredoxinproducts[index]\n",
    "                                    listferredoxins.append(SeqRecord(Seq(translationf),id=IDf))\n",
    "                        for index, value in enumerate(listp450locin):\n",
    "                                if value in listnewp450:\n",
    "                                    translation=listp450sequences[index]\n",
    "                                    locus=listp450loci[index]\n",
    "                                    ID=listp450proteins[index]\n",
    "                                    product=listp450products[index]\n",
    "                                    listp450.append(SeqRecord(Seq(translation),id=ID))\n",
    "                                    new_row={'Protein Reference':gi,'Acessiongenome':name1,\"p450 product\":product,\"p450 accessions\":ID,\"p450 loci\":str(locus),\"p450 translations\":translation,\"ferredoxin product\":productf,\"ferredoxin accession\":IDf,\"ferredoxin loci\":str(locusf),\"ferredoxin translation\":translationf}\n",
    "                                    tableofproteins = tableofproteins.append(new_row, ignore_index=True)\n",
    "                                if countsafe% 100 == 0:\n",
    "                                    tableofproteins.to_csv(filename_output, index=False) \n",
    "                                    SeqIO.write(listferredoxins, outputferre, 'fasta')\n",
    "                                    SeqIO.write(listp450, outputp450, 'fasta')\n",
    "              \n",
    "    except: pass\n",
    "    \n",
    "tableofproteins.to_csv(filename_output, index=False) \n",
    "SeqIO.write(listferredoxins, outputferre, 'fasta')\n",
    "SeqIO.write(listp450, outputp450, 'fasta')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
